src/snowflake/snowpark/_internal/compiler/plan_compiler.py

```diff
@@ -12,16 +12,48 @@
 
 
 class PlanCompiler:
-    """This class is common point of entry from compiling SnowflakePlan to list of queries and post actions.
-    We run pre-check and pre-process steps for optimization steps, and apply the activated optimizations.
+    """
+    This class is responsible for compiling a SnowflakePlan to list of queries and post actions that
+    will be sent over to the server for execution.
+
+    The entry point function is compile(), which applies the following steps:
+    1) Run pre-check for the Snowflake plan, which mainly checks if optimizations can be applied.
+    2) Run pre-process step if optimization can be applied, which extracts and copies the set of
+        logical plans associated with the original plan to apply optimizations on.
+    3) Applies steps of optimizations. Each optimization takes a set of logical plan and produces a
+        new set of logical plans. Note that the optimizations will not maintain schema/attributes,
+        so none of the optimizations should rely on the schema/attributes.
+    4) Generate queries.
     """
 
     def __init__(self, plan: SnowflakePlan) -> None:
         self._plan = plan
 
+    def should_apply_optimizations(self) -> bool:
+        """
+        Whether optimization should be applied to the plan or not.
+        Optimization can be applied if
+        1) there is source logical plan attached to the current snowflake plan
+        2) optimizations are enabled in the current session, such as cte_optimization_enabled
+
+
+        Returns
+        -------
+        True if optimization should be applied. Otherwise, return False.
+        """
+
+        current_session = self._plan.session
+        return (
+            self._plan.source_plan is not None
+        ) and current_session.cte_optimization_enabled
+
     def compile(self) -> Dict[PlanQueryType, List[Query]]:
-        # apply optimizations
-        final_plan = self._plan.replace_repeated_subquery_with_cte()
+        final_plan = self._plan
+        if self.should_apply_optimizations():
```

Comment:
shall we rename this to `should_apply_cte_optimization`. I'm assuming in the future it looks like

```
if should_apply_optimization_a():
    # apply optimization A
if should_apply_optimization_b():
    # apply optimization B
...
````

Reply:
eventually it will look like:
```
if self.should_apply_optimizations():
     pre-process()
     if optimization_a_enabled:
          optimization_A()
     if optimization_b_enabled:
         optimization_B()
    ....
    if optimization_triggered:
        generate_query
```